### ISSUE TITLE
Design Incident and Phenomena data model to support PQ classification

### ISSUE THREAD

**Issue by philipmjohnson** [openpowerquality/opq#121]
Please do work for this task in a branch called issue-121.

In [A beginners guide to power quality](https://openpowerquality.org/docs/intro-power-quality.html), a variety of PQ types are identified.  Investigate the creation of classification schemes for:

- Blackout
- Transient
- Over voltage
- Under voltage
- EMI
- RFI
- instantaneous
- momentary
- temporary
- swell
- prohibited region
- no interruption region
- no damage region

**Comment by philipmjohnson** [comment]
It's useful to do a google search for "power quality problem classification". For example, here's one article among several:

https://ieeexplore.ieee.org/document/5136965/

**Comment by anthonyjchriste** [comment]
Thanks, that's a great search term.

**Comment by anthonyjchriste** [comment]
After starting to delve into this a bit more, I'm of the opinion that our current data model for events may need to be updated. Events, as currently designed, are centered around triggering. That is, triggering sees something interesting, creates an "event" which in turn goes out and attempts to receive data which then creates "box_events".

This model doesn't provide a clear way for adding PQ classifications ex-post-facto once Mauka receives the raw waveform and starts performing higher level analysis. What Mauka does currently with THD and ITIC is it adds those values to fields within box_events, but I'm not convinced this is the best approach moving forward. It feels really icky.

I think we might be in need of a new collection, something like classified_[issues|events], where each document is a single classification rather than trying to aggregate all the PQ information into a single box_event.

Thoughts?

**Comment by philipmjohnson** [comment]
I agree, but I would take it even a little further.  How about this:

1. There is a new entity called "Anomaly" held in a collection called "anomalies"

2. Anomalies have these fields:
  * events: an array of Events associated with this anomaly.
  * type:  a string indicating the type of this anomaly
  * data: an object containing keys and values with additional information about this anomaly.

3. The event entity no longer has a "type" or "description" field.

4. The box_event entity no longer has "itic" or "thd" fields. (These become an Anomaly "type")

Basically, we remove all fields associated with "interpretation" of PQ data out of the event and box_event entities.  All they do is represent raw data.  All interpretation is now found in the Anomaly entity.  Anomalies are created by Mauka or Makai or whoever wants to look at events and box_events.

Note that an Anomaly can represent a periodic event and/or a problem that is found by looking at multiple events. For example, let's say that there's a voltage sag that occurs every day at noon, or every time the insolation drops by a specified amount in a very short period of time, or something.   The associated Mauka plugin could detect this, and then there would be a single Anomaly that represents this reoccurring event ("Voltage Sag At Noon") which incrementally augments its events field as new occurrences of voltage sags fitting that classification happen. 

How does this sound?

**Comment by anthonyjchriste** [comment]
I think it's a step in the correct direction. I'm not entirely sure I'm on board yet with an Anomaly representing more than one event. It would be nice to have a simple document that simply classifies every PQ issue we know how to classify as we see them. I think I would at least like to start building anomalies this way and then can maybe build them up to work with multiple events after we have them working with single PQ events.

**Comment by philipmjohnson** [comment]
I totally agree to start with "simple" Anomalies with a one-to-one correspondence with Events, but not sure why you want to have two different collections, one for "Simple Anomalies" and one for "Complex Anomalies".  It seems to me the only difference is whether the Events field takes a single Event or a list of Events.  What am I missing?

Looking at it again, I would replace the Events field by a "join table collection", which is a collection with two fields: AnomalyID and EventID.  That way, you can have fast bi-directional queries, i.e. you can easily look up all of the Events associated with an Anomaly, as well as all of the Anomalies associated with a given Event. Not sure why I didn't design it that way from the beginning.

Finally, now that I am looking at related work, I am somewhat doubtful that we will break any new ground through one-to-one Event to Anomaly classifications.  There are lots of very expensive PQ analyzers that do that already. For thesis-level innovation, I suspect the fertile ground is in classifications that are built off of multiple events. 

But, we have to crawl before we can walk. If you have a proposal for an alternative representation, or I am misunderstanding you, please reply!

**Comment by anthonyjchriste** [comment]
The branch should be issue-121 not issue-120.

**Comment by philipmjohnson** [comment]
Whoops! Fixed.

**Comment by philipmjohnson** [comment]
We're now thinking in terms of Incidents and Phenomena:

https://openpowerquality.org/docs/intro-information-architecture.html
